server/src/storage/uncompressed_data_manager.rs

```diff
@@ -47,23 +54,172 @@ impl UncompressedDataManager {
+                    Some((
+                        index,
+                        data_points.column(index as usize).as_any().downcast_ref().unwrap(),
+                    ))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // For each row in the data, generate a tag hash, extract the individual measurements,
+        // and insert them into the storage engine.
+        for (index, timestamp) in timestamps.iter().enumerate() {
+            let tag_values: Vec<String> = tag_column_arrays
+                .iter()
+                .map(|array| array.value(index).to_string())
+                .collect();
+
+            let tag_hash = self
+                .get_or_compute_tag_hash(&model_table, &tag_values)
+                .map_err(|error| format!("Tag hash could not be saved: {}", error.to_string()))?;
+
+            // For each field column, generate the 64-bit key, create a single data point, and
+            // insert the data point into the in-memory buffer.
+            for (field_index, field_column_array) in &field_column_arrays {
+                let key = tag_hash | *field_index as u64;
+                let value = field_column_array.value(index);
+
+                // unwrap() is safe to use since the timestamps array cannot contain null values.
+                self.insert_data_point(key, timestamp.unwrap(), value);
+            }
+        }
 
-        info!("Inserting data point '{}' into segment.", data_point);
+        Ok(())
+    }
+
+    /// Return the tag hash for the given list of tag values either by retrieving it from the cache
+    /// or, if it is a new combination of tag values, generating a new hash. If a new hash is
+    /// created, the hash is saved both in the cache and persisted to the model_table_tags table. If
+    /// the table could not be accessed, return [`rusqlite::Error`].
+    fn get_or_compute_tag_hash(
+        &mut self,
+        model_table: &NewModelTableMetadata,
+        tag_values: &Vec<String>,
+    ) -> Result<u64, rusqlite::Error> {
+        let cache_key = {
+            let mut cache_key_list = tag_values.clone();
+            cache_key_list.push(model_table.name.clone());
+
+            cache_key_list.join(";")
+        };
+
+        // Check if the tag hash is in the cache. If it is, retrieve it. If it is not, create a new
+        // one and save it both in the cache and in the model_table_tags table.
+        if let Some(tag_hash) = self.tag_value_hashes.get(&cache_key) {
+            Ok(*tag_hash)
+        } else {
+            // Generate the 54-bit tag hash based on the tag values of the record batch and model table name.
+            let tag_hash = {
+                let mut hasher = DefaultHasher::new();
+                hasher.write(cache_key.as_bytes());
+
+                // The 64-bit hash is shifted to make the 10 least significant bits 0.
+                hasher.finish() << 10
```

Comment:
use a constant?

Reply:
See [this](https://github.com/ModelarData/ModelarDB-RS/pull/30#discussion_r973097778) answer. In general I think making the key structure changeable is an interesting discussion. Especially considering the existence of potential customers that have somewhat ridiculous data requirements (more than 1024 columns). I have added a [work item](https://dev.azure.com/ModelarData/ModelarDB%20Implementation/_workitems/edit/261) to discuss this later but I think we should keep the code as it is for now.